### PR TITLE
feat(sql/auth): base authentication storage implementation

### PR DESCRIPTION
**Pull request by GeorgeMac**
Fixes: FLI-36

Initial implementation for storing `Authentication` instances into the backing relational DBs.
This PR:

1. Creates a new package `internal/storage/sql/auth`.
2. Creates an `*auth.Store` implementation specifically for `Authentication` associated state.
3. Adds the necessary migrations for each DB driver.
4. Implements `CreateAuthentication`.
5. Implements `GetAuthenticationByClientToken`.
6. Create and Get both perform client token hashing using `SHA256`.
7. Adds create and read tests for `Authentication` instances.

### CODE REVIEW

File: internal/storage/sql/errors.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	"go.flipt.io/flipt/internal/storage"
+)
+
+var (
+	errNotUnique          = fmt.Errorf("not unique: %w", storage.ErrInvalid)
+	errForeignKeyNotFound = fmt.Errorf("associated resource: %w", storage.ErrNotFound)
+)
+
+// AdaptError converts specific known-driver errors into wrapped storage errors.
+func (d Driver) AdaptError(err error) error {
```
Comment (markphelps) [inline internal/storage/sql/errors.go:20]: oh i like this. is the idea to replace the specific `sql/x` subpackages that override the `common` implementation with this for existing flag/segment/etc CRUD operations?

File: internal/storage/sql/auth/store.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+// NewStore constructs and configures a new instance of *Store.
+// Queries are issued to the database via the provided statement builder.
+func NewStore(driver fliptsql.Driver, builder sq.StatementBuilderType, logger *zap.Logger, opts ...Option) *Store {
+	store := &Store{
+		logger:  logger,
+		driver:  driver,
+		builder: builder,
+		now:     timestamppb.Now,
+		generateID: func() string {
+			return uuid.Must(uuid.NewV4()).String()
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/auth/store.go:46]: I think the only case `uuid` panics occur is when your system runs out of entropy (I could be mistaken). 
I believe the advice in this situation is to panic for reals. So, I wouldn't be too worried about not capturing this.

File: internal/storage/sql/auth/fields.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		return fmt.Errorf("unexpected method type: %T", v)
+	}
+
+	*m = method(auth.Method_value[methodStr])
+
+	return nil
+}
+
+func (m *method) Value() (driver.Value, error) {
+	return auth.Method_name[int32(*m)], nil
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/auth/fields.go:53]: Interesting, do they already support `database/sql` or are we storing them as integer types?

File: internal/storage/sql/auth/store.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+			"metadata",
+			"expires_at",
+			"created_at",
+			"updated_at",
+		).
+		Values(
+			&authentication.Id,
+			&hashedToken,
+			ptr(method(authentication.Method)),
+			&jsonField[map[string]string]{authentication.Metadata},
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/auth/store.go:120]: The `jsonField` container just marshals which type it contains to and from a string to make that operations a one-liner. I could definitely use `json.Compact` in the `Driver` implementation too. But this is the same thing.

File: internal/storage/sql/errors.go (hunk 1 above)
Comment (GeorgeMac) [reply internal/storage/sql/errors.go:20]: Yeah, 💯 I think we might be able to do away with having to add adaptors for each driver.
I can port the rest of Flipt to it in a later PR too if we like how it works.

File: internal/storage/sql/auth/store.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (markphelps) [reply internal/storage/sql/auth/store.go:120]: cool, i wonder if we should move to be consistent everywhere

File: internal/storage/sql/auth/fields.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/auth/fields.go:53]: I'll bring it in-line with the rest of Flipt. I was storing enums in their string form (i.e. `TOKEN`).
This was performing the mapping to and from the string form.
But the integer is going to potentially be more compact. So let's go with that.

File: internal/storage/sql/auth/store.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/auth/store.go]: Let me see if I can drop Attachment in, and it just works...
I just moved `JSONField` to the `sql` package, so it is reachable now.

[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/auth/store.go]: Ahh I see, Attachment moves around as a string in the protobuf itself. My mistake.

Whereas, Metadata is actually a `map<string, string>` in protobuf:
https://github.com/flipt-io/flipt/blob/35eec68a4347a536d2b711d47a105ccaef158b62/rpc/flipt/auth/auth.proto#L45

Here, `JSONField` allows for the one-lining of (un)marshalling to and from arbitrary Go types like this to string for persistence.

Whereas, `Attachment` using `json.Compact` to first unmarshal a string which assumed to contain JSON and back into a compact JSON form.

Seems these are two relatively different concepts, and I see why they're both different.
With Attachments, the data's schema is controlled by the end user.
We don't need to know the structure of the JSON document.
It is compacted to keep storage cost minimized.

With Metadata, the data's schema is strictly `map<string, string>`.
We use that structure internally to allow each auth method to populate the map in their respective ways (e.g. static tokens store their name and description under controlled keys).
JSON is merely a convenient serialization form to stash the structured object in a `TEXT` field.

File: internal/storage/sql/auth/store_test.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+				return
+			}
+
+			require.NoError(t, err)
+			assert.Equal(t, expectedAuthentication, retrieved)
+		})
+	}
+}
+
+func FuzzHashClientToken(f *testing.F) {
```
Comment (markphelps) [inline internal/storage/sql/auth/store_test.go:197]: ooo love the fuzz

**Review by markphelps — APPROVED** [review]
heyooooo 🚀

File: internal/storage/sql/auth/store_test.go (hunk 5 above)
Comment (GeorgeMac) [reply internal/storage/sql/auth/store_test.go:197]: Honestly, first time I have given it a proper go. It is really neat.
I ran a long fuzz of this function. Roughly ~7M calls.

It is mostly just base64 and sha256, so not testing much.
However, it was nice to get some confidence it was producing base64 encoded non-zero length tokens.

**Pull request merged**